canisters/src/vts/src/lib.rs

```diff
@@ -208,6 +215,79 @@ struct AgreementConditions {
     gas_price: String,
 }
 
+fn create_invoice(
+    vehicle_id: Principal,
+    start_period: String,
+    end_period: String,
+    aggregated_data: &AccumulatedTelemetry,
+) -> VTSResult<()> {
+    let existing_invoice = INVOICES.with(|invoices| {
+        let invoices = invoices.borrow();
+        invoices.iter().any(|invoice| {
+            invoice.1.vehicle == vehicle_id
+                && invoice.1.period.0 == start_period
+                && invoice.1.period.1 == end_period
+        })
+    });
+
+    if existing_invoice {
+        return Ok(());
+    }
+
+    let vehicle = VEHICLES
+        .with(|vehicles| {
+            let vehicles = vehicles.borrow();
+            vehicles.get(&vehicle_id)
+        })
+        .ok_or(Error::NotFound)?;
+
+    let agreement_conditions = vehicle
+        .agreement
+        .and_then(|agreement_id| {
+            AGREEMENTS.with(|agreements| {
+                let agreements = agreements.borrow();
+                agreements.get(&agreement_id).map(|agreement| agreement.conditions)
+            })
+        })
+        .ok_or(Error::NotFound)?;
+
+    let gas_price = Decimal::from_str(&agreement_conditions.gas_price).map_err(|_| Error::InvalidData)?;
+
+    let mut total_cost = Decimal::new(0, 0);
+
+    if let Some(aggregated_data) = aggregated_data.get(&TelemetryType::Gas) {
+        for usage in aggregated_data.values().map(|v| Decimal::new(v.value as i64, 0)) {
+            total_cost += usage * gas_price;
+        }
+    }
+
+    let invoice_id = INVOICE_ID_COUNTER.with(|counter| {
+        let mut counter = counter.borrow_mut();
+        *counter += 1;
+        *counter
+    });
+
+    let invoice = Invoice {
+        id: invoice_id,
+        vehicle: vehicle_id,
+        period: (start_period, end_period),
+        total_cost: total_cost.to_u64().ok_or(Error::InvalidData)?,
+    };
+
+    INVOICES.with(|invoices| invoices.borrow_mut().insert(invoice_id, invoice.clone()));
```

Comment:
Are you sure that you need to clone here? Because you do not use invoice below this line.